src/Orleans.Runtime/GrainDirectory/GrainDirectoryHandoffManager.cs

```diff
@@ -212,37 +208,79 @@ internal void ProcessSiloAddEvent(DirectoryMembershipSnapshot membershipSnapshot
                 // NOTE: We need to move part of our local directory to the new silo if it is an immediate successor.
                 var successor = membershipSnapshot.FindSuccessor(this.localSiloDetails.SiloAddress);
 
-                if (successor is null || !successor.Equals(addedSilo))
+                if (successor is null)
                 {
                     if (logger.IsEnabled(LogLevel.Debug)) logger.Debug($"{addedSilo} is not one of my successors.");
                     return;
                 }
 
-                // split my local directory and send to my new immediate successor his share
-                if (logger.IsEnabled(LogLevel.Debug)) logger.Debug("Splitting my partition between me and " + addedSilo);
-                GrainDirectoryPartition splitPart = localDirectory.DirectoryPartition.Split(
-                    grain =>
+                if (successor.Equals(addedSilo))
+                {
+                    // split my local directory and send to my new immediate successor his share
+                    if (logger.IsEnabled(LogLevel.Debug)) logger.Debug("Splitting my partition between me and " + addedSilo);
+                    GrainDirectoryPartition splitPart = localDirectory.DirectoryPartition.Split(
+                        grain =>
+                        {
+                            var s = membershipSnapshot.CalculateGrainDirectoryPartition(grain);
+                            return (s != null) && !this.localSiloDetails.SiloAddress.Equals(s);
```

Comment:
As a side note to interested readers, it's something to think how much work it would be to move from `SiloAddress` as in in `IPEndpoint` within to other ways to identify siloes.

Reply:
That would be great :)